tensorflow_datasets/image/coco_captions.py

```diff
@@ -0,0 +1,111 @@
+"""MSCOCO Captioning Dataset from 2014."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import json
+import tensorflow as tf
+import tensorflow_datasets.public_api as tfds
+
+_CITATION = """
+@article{DBLP:journals/corr/LinMBHPRDZ14,
+  author    = {Tsung{-}Yi Lin and
+               Michael Maire and
+               Serge J. Belongie and
+               Lubomir D. Bourdev and
+               Ross B. Girshick and
+               James Hays and
+               Pietro Perona and
+               Deva Ramanan and
+               Piotr Doll{\'{a}}r and
+               C. Lawrence Zitnick},
+  title     = {Microsoft {COCO:} Common Objects in Context},
+  journal   = {CoRR},
+  volume    = {abs/1405.0312},
+  year      = {2014},
+  url       = {http://arxiv.org/abs/1405.0312},
+  archivePrefix = {arXiv},
+  eprint    = {1405.0312},
+  timestamp = {Mon, 13 Aug 2018 16:48:13 +0200},
+  biburl    = {https://dblp.org/rec/bib/journals/corr/LinMBHPRDZ14},
+  bibsource = {dblp computer science bibliography, https://dblp.org}
+}
+"""
+
+_DESCRIPTION = """
+COCO is a large-scale object detection, segmentation, and captioning dataset.
+COCO has several features:
+
+Object segmentation
+Recognition in context
+Superpixel stuff segmentation
+330K images (>200K labeled)
+1.5 million object instances
+80 object categories
+91 stuff categories
+5 captions per image
+250,000 people with keypoints
+
+In this dataset, there are 414113 captions, but 82783 images. Each image-caption
+pair is represented once in the dataset - thus, each image appears 5 times.
+"""
+
+_HOMEPAGE = "http://cocodataset.org/"
+
+class CocoCaptions(tfds.core.GeneratorBasedBuilder):
+  """MSCOCO Image Captioning Dataset - 2014"""
+
+  VERSION = tfds.core.Version('1.0.0')
+
+  def _info(self):
+    return tfds.core.DatasetInfo(
+        builder=self,
+        description=_DESCRIPTION,
+        features=tfds.features.FeaturesDict({
+            'image': tfds.features.Image(encoding_format='jpeg'),
+            'caption': tfds.features.Text()
```

Comment:
I'd actually suggest instead of yielding multiple of the same image with different captions, yield one image with the 5 captions. Let me know if you need help implementing this.

Reply:
I see where this is coming from, but I think there's benefit in yielding them separately. While training, implementing the 5 captions-per-image will require the user to iterate through each caption, just adding an extra step. On the contrary, yielding them separately will allow the user to implement one train step per datapoint in the dataset.

Reply:
Hm, interesting. I was thinking that yielding multiple images would result in unnecessary overhead and possibly some issues when shuffling/batching the data. I guess we can open this up to some further discussion down thr road.

Reply:
For sure. We'll see what others think - by the way, I'll try to make the other changes you mentioned later today.